drivers/ethernet/eth_stm32_hal.c

```diff
@@ -1239,22 +1250,49 @@ static void net_if_mcast_cb(struct net_if *iface,
 	hash_table[0] = 0;
 	hash_table[1] = 0;
 
+	if (is_joined) {
+	/* Save a copy of the hash table which we update with
+	 * the hash for a single multicast address for join
+	 */
+#if defined(CONFIG_SOC_SERIES_STM32H7X)
+		hash_table[0] = heth->Instance->MACHT0R;
+		hash_table[1] = heth->Instance->MACHT1R;
+#else
+		hash_table[0] = heth->Instance->MACHTLR;
+		hash_table[1] = heth->Instance->MACHTHR;
+#endif /* CONFIG_SOC_SERIES_STM32H7X */
+	}
+
 #if defined(CONFIG_NET_NATIVE_IPV6)
 	if (net_if_config_ipv6_get(iface, &ipv6) < 0) {
 		return;
 	}
 
-	for (i = 0; i < NET_IF_MAX_IPV6_MADDR; i++) {
-		if (!ipv6->mcast[i].is_used) {
-			continue;
-		}
-
-		net_eth_ipv6_mcast_to_mac_addr(&ipv6->mcast[i].address.in_addr,
-					       &mac_addr);
+	if (is_joined) {
+		/* When joining only update the hash filter with the joining
+		 * multicast address.
+		 */
+		net_eth_ipv6_mcast_to_mac_addr(&addr->in6_addr, &mac_addr);
 		crc = reverse(crc32_ieee(mac_addr.addr,
-					 sizeof(struct net_eth_addr)));
+					sizeof(struct net_eth_addr)));
 		hash_index = (crc >> 26) & 0x3f;
 		hash_table[hash_index / 32] |= (1 << (hash_index % 32));
+	} else {
+		/* When leaving its better to compute the full hash table
+		 * for all the multicast addresses that we're aware of.
+		 */
+		for (i = 0; i < NET_IF_MAX_IPV6_MADDR; i++) {
+			if (!ipv6->mcast[i].is_used || !ipv6->mcast[i].is_joined) {
+				continue;
+			}
```

Comment:
Thanks @jukkar for the comment. I will kindly ask for a suggestion from you regarding what makes sense. The reason why the driver is accessing the multicast address list of the interface is to re-build the hash table in the case of a multicast leave. Essentially the multicast address list of the is the source of truth in terms of multicast addresses. If it doesn't make sense for the driver to access the multicast addresses of the interface do you think it would be best for the driver to maintain its own list of multicast addresses?

Reply:
My point is that for each joined multicast address, the upper stack will/should call this callback. So there should be no need to traverse the network interface list of multicast addresses. If this callback is not called for some reason, then we need to fix/change the upper stack code like you did for the the 224.0.01 address.

Reply:
Thanks @mrfuchs, yes I agree. The best way moving forward is to use a separate array in the driver. I'll be implementing this.

Reply:
Hi @jukkar, @mrfuchs, I've implemented a means in the driver to keep track of joined multicast addresses and use it for the hash calculation. Please review and let me know if there is anything that needs to be changed.